### PR TITLE
Concurrent migrations

### PR DESCRIPTION
**Pull request by shlomi-noach**
**See updated https://github.com/vitessio/vitess/pull/9192#issuecomment-968706668**

## Description

Ongoing work towards supporting concurrent REVERT migrations.

Up till now, Vitess would only ever run a single migration at a time, for a given keyspace.

- This isn't 100% correct, as a tablet failure could end up resurrecting an already vreplication stream.

Anyway, we wish to support the following:

- REVERT migrations, which are by nature high priority migrations, should be able to run concurrently to any other migration
- We will require an additional flag in `ddl_strategy`, something like `-allow-concurrent` or similar
- There can be an unlimited number of concurrent REVERT migrations with this flag
- But still only one single "normal" migration at a time.

This changes the scheduler logic, which up till now assumed that if _any_ migration was running, then no need to schedule the next one. From now on, it will need to consider these scenarios:

- if a migration is queued, it can be made ready if nothing else is running
- if a "allowed concurrent" migration is queued, it can be made ready is no other migration operates _on the same table_
- if any migration is queued, it can be made ready if only other running migrations are "allowed concurrent" and none of them operates on _same table_

This means a more programmatic approach to scheduling next migration (where today some of the logic is a simple SQL)

## Related Issue(s)

- This PR extends https://github.com/vitessio/vitess/pull/9171 (unmerged yet)
- tracking issue: #6926

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
While #9171 is still unmerged, this is the diff for this branch: https://github.com/planetscale/vitess/compare/online-ddl-postpone-completion...planetscale:online-ddl-concurrent-revert-migrations

**Comment by shlomi-noach** [comment]
This PR has evolved a bit. We now support concurrent migrations as follows:

- A migration needs to opt in to run concurrently via `-allow-concurrent` flag in `ddl_strategy`
- A migration is scheduled and executed concurrently if:
  - It specifies `-allow-concurrent`
  - Is `CREATE`, or `DROP`, or `REVERT` (`REVERT` works for `CREATE`, `DROP` and for `online`/vrplication `ALTER` migrations)
  - There is no other running migration operating on same table.

With the new scheduler:

- There can be as many concurrent migrations running at any given time
- And there can be at most one non-concurrent ("regular") migration running at any time
- And never two migrations operating on the same table

Also:

- There can be multiple migrations in `ready` state (previously only one)
- A migration can be "blocked" from running, if it conflicts with a running migration (operating on same table, or non-concurrent)
- While some `ready` migrations can be blocked, new migrations are still able to proceed to `ready` state an then to `running`, i.e. actually execute.

Just some possible scenarios:

- If all migrations are non-concurrent, i.e. are regular, then they can only run sequentially.
- If a regular migration is running, it's possible for a `-allow-concurrent` migration to also kick in and run
  - and then it's also possible for yet another N `-allow-concurrent` migrations to kick in and run
- If one or more `-allow-concurrent` migrations are running, it is possible for a regular migration to kick in and run
-

**Comment by shlomi-noach** [comment]
This is now ready to review, given https://github.com/vitessio/vitess/pull/9171 is merged.

File: go/vt/vttablet/onlineddl/executor.go [hunk 1]
```diff
@@ -2179,45 +2235,67 @@ func (e *Executor) executeMigration(ctx context.Context, onlineDDL *schema.Onlin
 	return nil
 }
 
+// runNextMigration picks one 'ready' migration that is able to run, and executes it.
```
Comment (shlomi-noach) [inline go/vt/vttablet/onlineddl/executor.go:2238]: This is at the heart of the PR: the scheduler here gets to decide whether there's a migration thta can be executed, concurrently or not concurrently, based on existence or non-existence of running migrations.

File: go/vt/vttablet/onlineddl/executor.go [hunk 2]
```diff
@@ -306,15 +308,86 @@ func (e *Executor) triggerNextCheckInterval() {
...
+		}
+		return true
 	})
-	return migrationFound
+	return sameTableMigrationFound
+}
+
+// isAnyNonConcurrentMigrationRunning sees if there's any migration running right now
+// that does not have -allow-concurrent.
+// such a running migration will for example prevent a new non-concurrent migration from running.
```
Comment (deepthi) [inline go/vt/vttablet/onlineddl/executor.go:380]: copy-pasta. Need the correct description of this function instead.

File: go/vt/vttablet/onlineddl/executor.go [hunk 3]
```diff
@@ -1561,36 +1634,15 @@ func (e *Executor) CancelPendingMigrations(ctx context.Context, message string)
 }
 
 // scheduleNextMigration attemps to schedule a single migration to run next.
-// possibly there's no migrations to run. Possibly there's a migration running right now,
-// in which cases nothing happens.
+// possibly there's no migrations to run.
```
Comment (deepthi) [inline go/vt/vttablet/onlineddl/executor.go:1637]: nit: there are

File: go/vt/vttablet/onlineddl/executor.go [hunk 4]
```diff
@@ -1561,36 +1634,15 @@ func (e *Executor) CancelPendingMigrations(ctx context.Context, message string)
...
+// The effect of this function is to move a migration from 'queued' state to 'ready' state, is all.
+// Notice that the query sqlScheduleSingleMigration embeds some logic inside. We may choose
+// to refactor the logic into the app
```
Comment (deepthi) [inline go/vt/vttablet/onlineddl/executor.go:1640]: hmm.. which app is this referring to?

File: go/vt/vttablet/onlineddl/executor.go [hunk 5]
```diff
@@ -2230,48 +2286,67 @@ func (e *Executor) executeMigration(ctx context.Context, onlineDDL *schema.Onlin
 	return nil
 }
 
+// runNextMigration picks one 'ready' migration that is able to run, and executes it.
+// Possible scenarios:
+// - no migration is in 'ready' state -- nothing to be done
+// - a migration is 'ready', but conflicts with other running migrations -- try another 'ready' migration
+// It is therefore possible that there is a 'ready' migration, and still this function only handles one.
```
Comment (deepthi) [inline go/vt/vttablet/onlineddl/executor.go:2293]: This sentence is confusing. I would have thought that it is possible that there is a "ready" migration and still this function chooses none.

File: go/vt/vttablet/onlineddl/executor.go [hunk 6]
```diff
@@ -2230,48 +2286,67 @@ func (e *Executor) executeMigration(ctx context.Context, onlineDDL *schema.Onlin
...
 
-	if e.isAnyMigrationRunning() {
-		return ErrExecutorMigrationAlreadyRunning
+	if !e.reviewedRunningMigrationsFlag {
+		// Since Open(), we havent's once executed reviewRunningMigrations() successfully.
+		// This means we may not have a good picture of what is actually running. Perhaps there's
+		// a vreplication migration from a pre-PRS/ERS that we still need to learn about?
+		// We're going to be careful here, and avoid running new migrations until we have
+		// a better picture. It will likely take a couple seconds till next iteration.
+		// execution. This delay ony takes place shortly after Open().
```
Comment (deepthi) [inline go/vt/vttablet/onlineddl/executor.go:2305]:
```suggestion
		// This delay only takes place shortly after Open().
```

File: go/vt/vttablet/onlineddl/executor.go [hunk 7]
```diff
@@ -2230,48 +2286,67 @@ func (e *Executor) executeMigration(ctx context.Context, onlineDDL *schema.Onlin
...
-			UUID:           row["migration_uuid"].ToString(),
-			Strategy:       schema.DDLStrategy(row["strategy"].ToString()),
-			Options:        row["options"].ToString(),
-			Status:         schema.OnlineDDLStatus(row["migration_status"].ToString()),
-			Retries:        row.AsInt64("retries", 0),
-			TabletAlias:    row["tablet"].ToString(),
-			RequestContext: row["migration_context"].ToString(),
+	// getNonConflictingMigration finds a single 'ready' migration which does not conflict with running migrations.
+	// Conflicts are:
+	// - a migration is 'ready' but is not set to run _concurrnetly_, and there's a running migration that is also non-concurrent
```
Comment (deepthi) [inline go/vt/vttablet/onlineddl/executor.go:2311]:
```suggestion
	// - a migration is 'ready' but is not set to run _concurrently_, and there's a running migration that is also non-concurrent
```

File: go/vt/vttablet/onlineddl/executor.go [hunk 8]
```diff
@@ -2230,48 +2286,67 @@ func (e *Executor) executeMigration(ctx context.Context, onlineDDL *schema.Onlin
...
+				return onlineDDL, err
 			}
 		}
-		e.executeMigration(ctx, onlineDDL)
-		// the query should only ever return a single row at the most
-		// but let's make it also explicit here that we only run a single migration
-		if i == 0 {
-			break
+		// no non-conflicting migration found...
+		// Either all ready migrations are conflicting, or there's no ready migrations...
```
Comment (deepthi) [inline go/vt/vttablet/onlineddl/executor.go:2330]: there's no ready migration OR there are no ready migrations.

**Review by deepthi — APPROVED** [review]
Feedback is all on code comments.
May merge after fixing them.

File: go/vt/vttablet/onlineddl/executor.go (hunk 4 above)
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go:1640]: Clarified, moved explanation into the function itself. This is more a developer's note for future me/us.

File: go/vt/vttablet/onlineddl/executor.go (hunk 5 above)
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go:2293]: Oh wow the wording was really bad. Changed to:

```
// Note that per the above breakdown, and due to potential conflicts, it is possible to have one or
// more 'ready' migration, and still none is executed.
```

**Comment by shlomi-noach** [comment]
Addressed all comments.

**Comment by shlomi-noach** [comment]
Woohoo! That's an important one merged.

**Pull request merged**